A90: Add List Method to gRPC Health Service (#468)
grpc/proposal: A##-list-health.md

> # Add List endpoint to gRPC Health service

Reviewer: It looks like the next unassigned number is A90.  Please rename this file to `A90-health-service-list-method.md` and change this line to the following:

```
A90: Add List Method to gRPC Health Service
----
```

> * **Author(s):** Marcos Huck

Reviewer: Please list this as a github login, preceded by an `@`.

> * **Author(s):** Marcos Huck.

Reviewer 2: Very minor NIT
```suggestion
* **Author(s):** Marcos Huck
```

Author: Appreciate it :)

* **Approver:** TBD
* **Status:** Draft  
* **Implemented in:** TBD
* **Last updated:** 2024-12-17  

> * **Discussion at:** TBD

Author: Done, will wait until it gets posted to update with the link.

## Abstract

> This proposal introduces a new `List` RPC endpoint for the Health service, allowing clients to retrieve the statuses of all monitored services or a filtered list of specific services. This feature simplifies integration with status-reporting dashboards and enhances observability for microservices.

Reviewer: What we're talking about adding here is a method, not an endpoint.  Please update terminology throughout.

## Background

> The existing Health service provides basic health check functionality but lacks a mechanism to retrieve a comprehensive list of all monitored servicesand their statuses. 

Author: ```suggestion
The existing Health service provides basic health check functionality but lacks a mechanism to retrieve a comprehensive list of all monitored services and their statuses. 
```

> The existing Health service provides basic health check functionality but lacks a mechanism to retrieve a comprehensive list of all monitored services and their statuses. 

Reviewer: One thing to keep in mind here is that the list of health services exposed by an application can change over the lifetime of the process.  New services can be added at any point, which means that the set of service names returned by the List method may change over the lifetime of the process.

Author: That seems reasonable, might be worth adding a note.

> This limitation makes it challenging for clients to aggregate health information across multiple services, particularly for use cases like publishing service statuses to dashboards such as [Cachet](https://cachethq.io/) or [Statuspage](https://www.atlassian.com/software/statuspage).

Reviewer: Can you help me understand the use-case here?  Both of those publishing services seem to be aimed at communicating overall service health to end-users.  However, the gRPC health service is really about exposing the health of an individual endpoint, not of the service as a whole.  In what context would these be used together?

Author: Thanks for the review. 

I imagine the following use case: 

We need to aggregate these endpoint statuses to get overall service health. 

We'd build an aggregator service that polls these endpoints, applies logic to determine service status, and then feeds that consolidated data to tools like `Cachet` or `Statuspage` for end-user visibility, and also for internal monitoring. So, while `gRPC` is endpoint-focused, the Health service is a key data source for a broader service health system. 

Adding a new `List` endpoint would allow towards that end.

> Kubernetes provides a similar capability with its `/readyz?verbose` endpoint, which lists the status of all components.This proposal aims to bring analogous functionality to the Health service, providing a unified view of service health.

Author: ```suggestion
Kubernetes provides a similar capability with its `/readyz?verbose` endpoint, which lists the status of all components. This proposal aims to bring analogous functionality to the Health service, providing a unified view of service health.
```

### Related Proposals

N/A

## Proposal

Introduce a new `List` RPC endpoint in the Health service with the following features:

- Retrieve the health statuses of all monitored services or a filtered list of specified services.
- Ensure the implementation is idempotent and side-effect free.
- Provide a clear schema for the request and response to facilitate integration with external tools.

### Proposed API Changes

https://github.com/grpc/grpc-proto/pull/143

### Temporary environment variable protection

N/A

## Rationale

Adding the List RPC endpoint to the Health service strikes a balance between functionality and simplicity.

### Alternative approaches
- Separate Service: Extract the new functionality into a standalone service. This avoids altering the existing Health service API but increases complexity by introducing another service for clients to interact with.

## Implementation

1. Add the `List` RPC endpoint and the associated request/response messages in the Health service .proto file.

> 2. Update client libraries to support the `List` endpoint. Provide examples demonstrating how to call the endpoint and handle its response.

Reviewer: I think there's another important step before this one, which is to implement the new method in the health service implementations in each gRPC language.

Author: Would you mind pointing me in the right direction? I'm happy to write code in pretty much all the languages supported by gRPC, but I just want to make sure what are those places I need to change.

Reviewer 3: In Node the health service is implemented here: https://github.com/grpc/grpc-node/blob/master/packages/grpc-health-check/src/health.ts#L82. The `List` method would need to be added. It should mostly just be straightforwardly iterating through the `statusMap`.

Author: Thank you all for the guidance.
